src/v/resource_mgmt/cpu_profiler.cc

```diff
@@ -146,4 +149,33 @@ void cpu_profiler::on_sample_period_change() {
     ss::engine().set_cpu_profiler_period(_sample_period());
 }
 
+ss::future<std::vector<cpu_profiler::shard_samples>>
+cpu_profiler::override_and_get_results(
+  std::chrono::milliseconds timeout, std::optional<ss::shard_id> shard_id) {
+    if (_gate.is_closed()) {
+        co_return std::vector<shard_samples>{};
+    }
+    auto holder = _gate.hold();
+
+    auto currently_enabled = is_enabled();
+    _override_enabled++;
+
+    // If there are no other enable overrides currently then signal an enable
+    // change.
+    if (!currently_enabled) {
+        on_enabled_change();
+    }
+
+    co_await ss::sleep(timeout);
+    _override_enabled--;
+
+    // If other enable overrides are still on-going don't signal an enable
+    // change.
+    if (!is_enabled()) {
+        on_enabled_change();
```

Comment:
I feel like it would be simpler if we could just call `on_enabled_change()` unconditionally anywhere and inside there it would determine if anything needs to happen: i.e, if the seastar level enable != the repdanda enable, rather than having to carefully detect the cases where enablement may have changed in either direction and call it unconditionally.

That said, I think the logic here is correct.

Reply:
Calling it without any checks would work as well given the logic of `on_enabled_change()`. This was probably a premature optimization to avoid the unneeded system calls associated with re-enabling the profiler.

Reply:
> Calling it without any checks would work as well given the logic of `on_enabled_change()`. This was probably a premature optimization to avoid the unneeded system calls associated with re-enabling the profiler.

Right, I noticed that, though the optimization you point out might be a worthwhile optimization, but it could be done with 1 check inside `on_enabled_change` which just checks if the states are out of sync.

Reply:
Good point, switching to checking out-of-sync states within `on_enable_change` rather than at the locations its called.